Source/Core/DolphinQt/MemoryEngine/MemWatcher/Dialogs/DlgAddWatchEntry.cpp

```diff
@@ -0,0 +1,444 @@
+  if (m_entry->getPointerLevel() == 1)
+  {
+    m_btnRemoveOffset->setDisabled(true);
+    m_btnAddOffset->setFocus();
+  }
+}
+
+void DlgAddWatchEntry::removeAllPointerOffset()
+{
+  int level = static_cast<int>(m_entry->getPointerLevel());
+  while (level != 0)
+  {
+    QLayoutItem* lblLevelItem = m_offsetsLayout->takeAt(
+        m_offsetsLayout->indexOf(m_offsetsLayout->itemAtPosition(level - 1, 0)->widget()));
+    QLayoutItem* txbOffsetItem = m_offsetsLayout->takeAt(
+        m_offsetsLayout->indexOf(m_offsetsLayout->itemAtPosition(level - 1, 1)->widget()));
+    QLayoutItem* lblAddressOfPathItem = m_offsetsLayout->takeAt(
+        m_offsetsLayout->indexOf(m_offsetsLayout->itemAtPosition(level - 1, 2)->widget()));
+
+    delete lblLevelItem->widget();
+    delete txbOffsetItem->widget();
+    delete lblAddressOfPathItem->widget();
+
+    m_offsets.removeLast();
+    m_addressPath.removeLast();
+    m_entry->removeOffset();
+
+    level--;
+  }
+
+  updatePreview();
+}
+
+void DlgAddWatchEntry::onOffsetChanged()
+{
+  QLineEdit* theLineEdit = static_cast<QLineEdit*>(sender());
+  int index = 0;
+  // Dummy variable for getItemPosition
+  int column, rowSpan, columnSpan;
+  m_offsetsLayout->getItemPosition(m_offsetsLayout->indexOf(theLineEdit), &index, &column, &rowSpan,
+                                   &columnSpan);
+  if (validateAndSetOffset(index))
+    updatePreview();
+}
+
+void DlgAddWatchEntry::onTypeChange(int index)
+{
+  Common::MemType theType = static_cast<Common::MemType>(index);
+  if (theType == Common::MemType::type_string || theType == Common::MemType::type_byteArray)
+    m_lengtWidget->show();
+  else
+    m_lengtWidget->hide();
+  m_entry->setTypeAndLength(theType, m_spnLength->value());
+  if (validateAndSetAddress())
+    updatePreview();
+}
+
+void DlgAddWatchEntry::accept()
+{
+  if (!validateAndSetAddress())
+  {
+    QString errorMsg =
+        tr("The address you entered is invalid, make sure it is an "
+           "hexadecimal number between 0x80000000 and 0x817FFFFF");
+    if (Common::getMEM2())
+      errorMsg.append(tr(" or between 0x90000000 and 0x93FFFFFF"));
+
+    QMessageBox* errorBox = new QMessageBox(QMessageBox::Critical, tr("Invalid address"), errorMsg,
+                                            QMessageBox::Ok, this);
+    errorBox->exec();
+  }
+  else
+  {
+    if (m_chkBoundToPointer->isChecked())
+    {
+      bool allOffsetsValid = true;
+      int i = 0;
+      for (i; i < m_offsets.count(); ++i)
+      {
+        allOffsetsValid = validateAndSetOffset(i);
+        if (!allOffsetsValid)
+          break;
+      }
+      if (!allOffsetsValid)
+      {
+        QMessageBox* errorBox = new QMessageBox(
+            QMessageBox::Critical, tr("Invalid offset"),
+            QString::fromStdString("The offset you entered for the level " + std::to_string(i + 1) +
+                                   " is invalid, make sure it is an "
+                                   "hexadecimal number"),
```

Comment:
This should probably use `tr(...)` as well, and insert `i + 1` using a placeholder. Otherwise this string is left untranslatable compared to everything else in here.